trestle/core/remote/cache.py

```diff
@@ -0,0 +1,487 @@
+                            abbreviatedOid
+                            id
+                            isBinary
+                            isTruncated
+                            oid
+                            text
+                        }
+                    }
+                }
+            }
+            """
+        self._variables = {'owner': owner, 'name': name, 'rev': rev + ':' + str(src_filepath)}
+
+    def _sync_cache(self) -> None:
+        request = requests.post(
+            self._api,
+            json={
+                'query': self._query, 'variables': self._variables
+            },
+            headers={'Authorization': 'Bearer ' + self._token}
+        )
+        if request.status_code == 200:
+            result = request.json()
+            result = result['data']['repository']['object']
+            if result is None:
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), str(self._inst_cache_path))
+            if result['isBinary']:
+                raise NotImplementedError('Binary files are not supported!')
+            else:
+                self._inst_cache_path.write_text(result['text'])
+        else:
+            raise TrestleError(f'Query failed to run by returning code of {request.status_code}. {self._query}')
+
+
+class FetcherFactory(object):
+    """Factory method for creating a fetcher."""
+
+    def __init__(self) -> None:
+        """Initialize fetcher factory."""
+        pass
+
+    @classmethod
+    def get_fetcher(
+        cls,
+        trestle_root: pathlib.Path,
+        uri: str,
+        settings: Settings,
+        refresh: bool = False,
+        fail_hard: bool = False,
+        cache_only: bool = False
+    ) -> FetcherBase:
+        """Return an instantiated fetcher object based on the uri."""
+        # Basic correctness test
+        if len(uri) <= 9 or ('/' not in uri and 'C:\\' not in uri):
+            raise TrestleError(f'Unable to fetch uri as it appears to be invalid {uri}')
+
+        if uri[0] == '/' or uri[0:3] == '../' or uri[0:2] == './' or 'file:///' == uri[0:8]:
+            # Note assumption here is that relative paths are only supported within
+            # trestle directories. This simplification is to ensure
+            return LocalFetcher(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+        elif 'sftp://' == uri[0:7]:
+            return SFTPFetcher(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+        elif 'https://' == uri[0:8]:
+            # Test for github uri assumption - must be first after basic auth (if it exists)
+            cleaned = uri[8:]
+            # tests for special scenarios
+            if cleaned.split('@')[-1][0:7] == 'github.':
+                return GithubFetcher(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+            else:
+                return HTTPSFetcher(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+        elif 'C:\\' == uri[0:3]:
```

Comment:
Shouldn't this support any drive letter?

Reply:
Done.